redisx/cacher.go

```diff
@@ -0,0 +1,432 @@
+		default:
+			event, err := redis.Values(conn.Receive())
+			if err != nil {
+				continue outer
+			}
+
+			etype, _ := redis.String(event[0], nil)
+
+			if etype != "message" {
+				continue outer
+			}
+
+			keys, err := redis.Strings(event[2], nil)
+			if err != nil || err == redis.ErrNil {
+				continue outer
+			}
+
+			invalidateChan <- keys
+		}
+	}
+
+	close(invalidateChan)
+}
+
+// changed in tests to avoid sleeping insane amounts of time
+var pingerInterval = time.Second
+
+// pinger will try to ping server and will send to channel on multiple failed attempts
+func pinger(ctx context.Context, conn redis.Conn, connDeadChan chan<- struct{}) {
+	ticker := time.NewTicker(pingerInterval)
+	failures := 0
+
+outer:
+	for {
+		select {
+		case <-ctx.Done():
+			break outer
+		case <-ticker.C:
+			// Check connection is still active
+			if _, err := conn.Do("PING"); err == nil {
+				failures = 0
+				continue
+			}
+
+			failures++
+
+			if failures == 3 {
+				connDeadChan <- struct{}{}
+				failures = 0
+			}
+		}
+	}
+
+	ticker.Stop()
+	close(connDeadChan)
+}
+
+// Get returns a connection with caching enabled for matching keys.
+// If nil Matcher is passed all keys are subjected to cache.
+// When using Broadcast strategy Matcher should be created with NewPrefixMatcher.
+// Caller is responsible for closing connection.
+func (c *Cacher) Get(m Matcher) redis.Conn {
+	if c.Getter == nil {
+		panic(errNilGetter)
+	}
+	return c.Wrap(c.Getter.Get(), m)
+}
+
+var errNotRunning = errors.New("redisx cacher: invalidation process is not running")
+var errPrefixMatcherExpected = errors.New("redisx cacher: NewPrefixMatcher is expected when using Broadcast strategy")
+
+// Wrap allows wrapping existing connection with caching layer
+// When using Broadcast strategy Matcher should be created with NewPrefixMatcher.
+// Caller is responsible for closing connection.
+func (c *Cacher) Wrap(conn redis.Conn, m Matcher) redis.Conn {
+	c.mu.Lock()
+	cid := c.cid
+	c.mu.Unlock()
+
+	if cid == 0 {
+		panic(errNotRunning)
+	}
+
+	args := []interface{}{"TRACKING", "on", "REDIRECT", cid}
+
+	// When using Broadcast strategy we just pass list of key prefixes
+	// we are interested in.
+	if c.Strategy == Broadcast {
+		prefixes, ok := m.(prefixMatcher)
+		if !ok {
+			panic(errPrefixMatcherExpected)
+		}
+
+		args = append(args, "BCAST")
+		for _, prefix := range prefixes {
+			args = append(args, "PREFIX", prefix)
+		}
+	}
+
+	// If matcher if supplied we want to pass OPTIN flag so we can munually
+	// decide whether we want to cache certain key
+	if c.Strategy == Tracking && m != nil {
+		args = append(args, "OPTIN")
+	}
+
+	_, err := conn.Do("CLIENT", args...)
+	if err != nil {
+		return errorConn{err}
+	}
+
+	return wrappedConn{c, conn, m}
+}
+
+// Stats returns aggregated stats about local cache
+func (c *Cacher) Stats() CacheStats {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	size, hits, misses := c.cache.Stats()
+	return CacheStats{size, hits, misses}
+}
```

Comment:
```suggestion
func (c *Cacher) Stats() CacheStats {
	c.mu.Lock()
	size, hits, misses := c.cache.Stats()
	c.mu.Unlock()
	return CacheStats{size, hits, misses}
}
```

Should probably avoid deferring operations on the mutex in a cache, since deferring adds an extra ~200ns of latency.
Normally, 200ns would be extremely negligible, but for a cache, 200ns matters.

Reply:
It's actually like 4ns difference. A quick search reveals that defer performance has been [improved](https://golang.org/doc/go1.14#runtime) in Go 1.14 . I guess there was some overhead before that version.
Also no allocation happens here so everything runs as fast as it can get.

Reply:
My bad, didn't know about that.  Yea, 4ns is definitely negligible then.